### PR TITLE
feat(condo): DOMA-9348 ticket document generation

### PR DESCRIPTION
**Pull request by Alllex202**
- added a task for generating documents for ticket
At the moment, it is only possible to generate a document on completed work in the `docx` format
- removed icons from buttons on action bar on ticket page
- updated component `TaskProgress` (now, if all tasks are completed, then there is only one common cross to close the tasks)

### CODE REVIEW

File: apps/condo/domains/ticket/tasks/generateTicketDocument.js [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+        task = await TicketDocumentGenerationTask.getOne(context, {
+            id: taskId,
+            deletedAt: null,
+        })
+        if (!task) throw new Error('not found task!')
+
+        const ticketId = get(task, 'ticket.id')
+        if (!ticketId) throw new Error('not found ticketId!')
+        const ticket = await getById('Ticket', task.ticket.id)
+        if (!ticket || ticket.deletedAt) throw new Error('not found ticket!')
```
Comment (nomerdvadcatpyat) [inline apps/condo/domains/ticket/tasks/generateTicketDocument.js:44]: add taskId in logs

File: apps/condo/domains/ticket/utils/serverSchema/TicketDocumentGeneration/completionWorks.js [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+
+    const listOfWorks = invoices.reduce((acc, invoice) => {
+        const rows = Array.isArray(invoice.rows) ? invoice.rows : []
+        acc.push(...rows.map(row => {
+            const price = parseFloat(row.toPay)
+
+            return {
+                name: row.name || '',
+                count: String(row.count) || '',
+                price: !Number.isNaN(price) ? renderMoney(price, currencyCode, locale) : '',
```
Comment (nomerdvadcatpyat) [inline apps/condo/domains/ticket/utils/serverSchema/TicketDocumentGeneration/completionWorks.js:87]: If `price` === 0 and `isMin` = true it's mean that it's a `Contract` price. Probably, you need to support it here

File: apps/condo/domains/ticket/schema/TicketDocumentGenerationTask.test.js [hunk 3]
```diff
@@ -0,0 +1,451 @@
...
+                    format: TICKET_DOCUMENT_GENERATION_TASK_FORMAT.DOCX,
+                    documentType: TICKET_DOCUMENT_TYPE.COMPLETION_WORKS,
+                })
+                const tasks = await TicketDocumentGenerationTask.getAll(staff, {
+                    id: taskWithOtherUser.id,
+                })
+                expect(tasks).toHaveLength(0)
+            })
+
+            test('can update its task', async () => {
```
[the remark this answers is not in the document]
Comment (Alllex202) [reply apps/condo/domains/ticket/schema/TicketDocumentGenerationTask.test.js]: added test

File: apps/condo/domains/ticket/tasks/generateTicketDocument.js (hunk 1 above)
Comment (Alllex202) [reply apps/condo/domains/ticket/tasks/generateTicketDocument.js:50]: added logger

File: apps/condo/domains/ticket/utils/serverSchema/TicketDocumentGeneration/completionWorks.js (hunk 2 above)
Comment (Alllex202) [reply apps/condo/domains/ticket/utils/serverSchema/TicketDocumentGeneration/completionWorks.js:88]: fixed. if `isMin` = `true` then print nothing

**Review by nomerdvadcatpyat — APPROVED** [review]

File: apps/condo/domains/common/utils/createExportFile.js [hunk 4]
```diff
@@ -25,7 +39,9 @@ const render = (pathToTemplate, replaces, options = {}) => new Promise((resolve,
 // This function uses file saving apart from internal working of `File` field in Keystone schema
 // It makes unable to save data to `File` field server-side
 // @deprecated use `buildExportFile` like in `apps/condo/domains/ticket/tasks/exportTicketsTask.js`
-async function createExportFile ({ fileName, templatePath, replaces, meta }) {
+async function createExportFile ({ fileName, templatePath, replaces, meta, format = EXCEL }) {
```
Comment (toplenboren) [inline apps/condo/domains/common/utils/createExportFile.js:42]: best to leave hints that it defaults to excel

Comment (toplenboren) [reply apps/condo/domains/common/utils/createExportFile.js:42]: why do we use it, it is deprecated

File: apps/condo/domains/ticket/schema/TicketDocumentGenerationTask.js [hunk 5]
```diff
@@ -0,0 +1,203 @@
...
+
+const { getFileMetaAfterChange } = FileAdapter
+
+
+const TICKET_DOCUMENT_GENERATION_TASK_FOLDER_NAME = 'TicketDocumentGenerationTask'
+const TicketDocumentGenerationTaskFileAdapter = new FileAdapter(TICKET_DOCUMENT_GENERATION_TASK_FOLDER_NAME)
+const setFileMetaAfterChange = getFileMetaAfterChange(TicketDocumentGenerationTaskFileAdapter, 'file')
+
+const ERRORS = {
+    STATUS_IS_ALREADY_COMPLETED: {
```
Comment (toplenboren) [inline apps/condo/domains/ticket/schema/TicketDocumentGenerationTask.js:30]: best to make it 1 error

File: apps/condo/domains/ticket/schema/TicketDocumentGenerationTask.js [hunk 6]
```diff
@@ -0,0 +1,203 @@
...
+            kmigratorOptions: { null: false, on_delete: 'models.CASCADE' },
+            access: {
+                read: true,
+                create: true,
+                update: false,
+            },
+        },
+
+        timeZone: {
+            schemaDoc: 'To requested timeZone all datetime fields will be converted',
```
Comment (toplenboren) [inline apps/condo/domains/ticket/schema/TicketDocumentGenerationTask.js:118]:
```suggestion
            schemaDoc: 'All datetime fields in export template will be converted to specified tz',
```

File: apps/condo/domains/ticket/tasks/generateTicketDocument.js [hunk 7]
```diff
@@ -0,0 +1,107 @@
...
+
+        const organization = await getById('Organization', ticket.organization)
+        if (!organization || organization.deletedAt) throw new Error('not found organization!')
+
+        const country = get(organization, 'country', conf.DEFAULT_LOCALE)
+        const locale = get(COUNTRIES, country).locale
+        const documentType = get(task, 'documentType')
+
+        if (!(get(AVAILABILITY_TICKET_DOCUMENT, [documentType, 'supportedLocales']) || []).includes(locale)) {
```
Comment (toplenboren) [inline apps/condo/domains/ticket/tasks/generateTicketDocument.js:59]:
```suggestion
        if (!(get(SUPPORTED_DOCUMENT_TYPES_BY_LOCALE, [documentType, 'supportedLocales']) || []).includes(locale)) {
```

File: apps/condo/domains/ticket/tasks/generateTicketDocument.js [hunk 8]
```diff
@@ -0,0 +1,107 @@
...
+            throw new Error(`unsupported locale "${locale}" for documentType "${documentType}"`)
+        }
+
+        setLocaleForKeystoneContext(context, locale)
+
+        let fileUploadInput
+
+        switch (documentType) {
+            case TICKET_DOCUMENT_TYPE.COMPLETION_WORKS: {
```
Comment (toplenboren) [inline apps/condo/domains/ticket/tasks/generateTicketDocument.js:68]: maybe worksheet

File: apps/condo/domains/ticket/tasks/generateTicketDocument.js [hunk 9]
```diff
@@ -0,0 +1,107 @@
...
+                fileUploadInput = await generateTicketDocumentOfCompletionWorks({ task, locale, organization, ticket, context, baseAttrs: BASE_ATTRS })
+                break
+            }
+
+            default: {
+                throw new Error(`unexpected document type "${documentType}" for a document generations`)
```
Comment (toplenboren) [inline apps/condo/domains/ticket/tasks/generateTicketDocument.js:74]:
```suggestion
                throw new Error(`unexpected document type "${documentType}" for a document generation`)
```

File: apps/condo/domains/ticket/tasks/generateTicketDocument.js [hunk 10]
```diff
@@ -0,0 +1,107 @@
...
+            }
+        }
+
+        if (!fileUploadInput) throw new Error('not generated document for uploading!')
```
Comment (toplenboren) [inline apps/condo/domains/ticket/tasks/generateTicketDocument.js:78]:
```suggestion
        if (!fileUploadInput) throw new Error('failed to generate document for uploading!')
```

File: apps/condo/domains/ticket/tasks/generateTicketDocument.js [hunk 11]
```diff
@@ -0,0 +1,107 @@
...
+            await TicketDocumentGenerationTask.update(context, task.id, {
+                ...BASE_ATTRS,
+                status: TICKET_DOCUMENT_GENERATION_TASK_STATUS.ERROR,
+                meta: { ...task.meta, error: error.message },
+            })
+        }
+
+        throw error
+    }
+
```
Comment (toplenboren) [inline apps/condo/domains/ticket/tasks/generateTicketDocument.js:101]:
```suggestion
```

File: apps/condo/domains/ticket/utils/serverSchema/TicketDocumentGeneration/completionWorks.js [hunk 12]
```diff
@@ -0,0 +1,144 @@
...
+}
+
+const generateTicketDocumentOfCompletionWorks = async ({ task, baseAttrs, context, locale, ticket, organization }) => {
+    const { format, timeZone: timeZoneFromUser } = task
+
+    const timeZone = normalizeTimeZone(timeZoneFromUser) || DEFAULT_ORGANIZATION_TIMEZONE
+    const printDate = dayjs().tz(timeZone).locale(locale)
+
+    const property = await getById('Property', ticket.property)
+    const { renderData, streetPart, settlement, housePart, cityType, cityName } = getAddressDetails(get(property, 'addressMeta', ticket.propertyAddressMeta))
```
Comment (toplenboren) [inline apps/condo/domains/ticket/utils/serverSchema/TicketDocumentGeneration/completionWorks.js:55]: what if no address details

File: apps/condo/lang/en/en.json [hunk 13]
```diff
@@ -912,6 +912,8 @@
   "pages.condo.ticket.exportBlank.HaveAllComments.tooltip": "If multiple tickets are selected, all comments will be printed. You won't be able to select specific ones.",
   "pages.condo.ticket.exportBlank.ShowAllComments.label": "Show comments",
   "pages.condo.ticket.exportBlank.HideAllComments.label": "Hide comments",
+  "pages.condo.ticket.generateDocument.label": "Generate document",
+  "pages.condo.ticket.generateDocument.completionWorks.label": "Completion works",
```
Comment (toplenboren) [inline apps/condo/lang/en/en.json:916]: Worksheet

Comment (toplenboren) [reply apps/condo/lang/en/en.json:916]: or Work completion certificate

File: apps/condo/lang/ru/ru.json [hunk 14]
```diff
@@ -912,6 +912,8 @@
   "pages.condo.ticket.exportBlank.HaveAllComments.tooltip": "Если выбрано несколько заявок — в печать уйдут все комментарии. Выбрать конкретные не получится.",
   "pages.condo.ticket.exportBlank.ShowAllComments.label": "Показать комментарии",
   "pages.condo.ticket.exportBlank.HideAllComments.label": "Скрыть комментарии",
+  "pages.condo.ticket.generateDocument.label": "Сформировать акт",
+  "pages.condo.ticket.generateDocument.completionWorks.label": "Акт выполненных работы",
```
Comment (toplenboren) [inline apps/condo/lang/ru/ru.json:916]:
```suggestion
  "pages.condo.ticket.generateDocument.completionWorks.label.plural": "Акт выполненных работы",
```

File: apps/condo/migrations/20240628174659-0405_ticketdocumentgenerationtaskhistoryrecord_and_more.js [hunk 15]
```diff
@@ -0,0 +1,46 @@
...
+ALTER TABLE "TicketDocumentGenerationTask" ADD CONSTRAINT "TicketDocumentGenerationTask_ticket_3adfa397_fk_Ticket_id" FOREIGN KEY ("ticket") REFERENCES "Ticket" ("id") DEFERRABLE INITIALLY DEFERRED;
+ALTER TABLE "TicketDocumentGenerationTask" ADD CONSTRAINT "TicketDocumentGenerationTask_updatedBy_6fa8e411_fk_User_id" FOREIGN KEY ("updatedBy") REFERENCES "User" ("id") DEFERRABLE INITIALLY DEFERRED;
+ALTER TABLE "TicketDocumentGenerationTask" ADD CONSTRAINT "TicketDocumentGenerationTask_user_91522098_fk_User_id" FOREIGN KEY ("user") REFERENCES "User" ("id") DEFERRABLE INITIALLY DEFERRED;
+CREATE INDEX "TicketDocumentGenerationTask_createdAt_3ddb4fc2" ON "TicketDocumentGenerationTask" ("createdAt");
+CREATE INDEX "TicketDocumentGenerationTask_updatedAt_4f704bf6" ON "TicketDocumentGenerationTask" ("updatedAt");
+CREATE INDEX "TicketDocumentGenerationTask_deletedAt_165d0326" ON "TicketDocumentGenerationTask" ("deletedAt");
+CREATE INDEX "TicketDocumentGenerationTask_createdBy_2a072a7d" ON "TicketDocumentGenerationTask" ("createdBy");
+CREATE INDEX "TicketDocumentGenerationTask_ticket_3adfa397" ON "TicketDocumentGenerationTask" ("ticket");
```
Comment (toplenboren) [inline apps/condo/migrations/20240628174659-0405_ticketdocumentgenerationtaskhistoryrecord_and_more.js:24]: maybe add status index

**Review by toplenboren — APPROVED** [review]

**Pull request merged**